migrate/main.go

```diff
@@ -0,0 +1,303 @@
+	cmd.PersistentFlags().StringVar(&clusterManifest, "cluster-manifest", "", "path to public manifest backup")
+	cmd.PersistentFlags().StringVar(&outputDir, "output-dir", "./manifests", "manifest output directory")
+	_ = cmd.MarkPersistentFlagRequired("cloud-provider")
+	_ = cmd.MarkPersistentFlagRequired("cloud-region")
+	_ = cmd.MarkPersistentFlagRequired("crdb-cluster")
+
+	return cmd
+}
+
+func To[T any](v T) *T {
+	return &v
+}
+
+func yamlToDisk(path string, data any) error {
+	file, err := os.Create(path)
+	if err != nil {
+		return errors.Wrap(err, "creating file")
+	}
+	bytes, err := yaml.Marshal(data)
+	if err != nil {
+		return errors.Wrap(err, "marshalling yaml")
+	}
+	// Hack: drop creationTimestamp: null lines. See
+	// https://github.com/kubernetes/kubernetes/issues/67610 for details.
+	lines := strings.Split(string(bytes), "\n")
+	filteredLines := []string{}
+	timestampRE := regexp.MustCompile(`\s*creationTimestamp: null`)
+	for _, line := range lines {
+		if !timestampRE.MatchString(line) {
+			filteredLines = append(filteredLines, line)
+		}
+	}
+	if _, err := file.Write([]byte(strings.Join(filteredLines, "\n"))); err != nil {
+		return errors.Wrap(err, "writing yaml")
+	}
+	return nil
+}
+
+func buildNodeSpec(cluster publicv1.CrdbCluster, sts *appsv1.StatefulSet, nodeName string, joinString string, flags map[string]string) v1alpha1.CrdbNodeSpec {
+	return v1alpha1.CrdbNodeSpec{
+		NodeName:  nodeName,
+		Join:      joinString,
+		PodLabels: sts.Spec.Template.Labels,
+		Flags:     flags,
+		DataStore: v1alpha1.DataStore{
+			VolumeClaimTemplate: &corev1.PersistentVolumeClaim{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "datadir",
+				},
+				Spec: sts.Spec.VolumeClaimTemplates[0].Spec,
+			},
+		},
+		Domain:               "",
+		LoggingConfigMapName: cluster.Spec.LogConfigMap,
+		Env: append(cluster.Spec.PodEnvVariables, []corev1.EnvVar{
+			{
+				Name: "HostIP",
+				ValueFrom: &corev1.EnvVarSource{
+					FieldRef: &corev1.ObjectFieldSelector{
+						APIVersion: "v1",
+						FieldPath:  "status.hostIP",
+					},
+				},
+			},
+		}...),
+		ResourceRequirements: sts.Spec.Template.Spec.Containers[0].Resources,
+		Image:                sts.Spec.Template.Spec.Containers[0].Image,
+		ServiceAccountName:   "cockroachdb",
+		GRPCPort:             cluster.Spec.GRPCPort,
+		SQLPort:              cluster.Spec.SQLPort,
+		HTTPPort:             cluster.Spec.HTTPPort,
+		Certificates: v1alpha1.Certificates{
+			ExternalCertificates: &v1alpha1.ExternalCertificates{
+				CAConfigMapName:         cluster.Name + "-ca",
+				NodeSecretName:          cluster.Name + "-node-certs",
+				RootSQLClientSecretName: cluster.Name + "-client-certs",
+			},
+		},
+		Affinity:               cluster.Spec.Affinity,
+		NodeSelector:           cluster.Spec.NodeSelector,
+		Tolerations:            cluster.Spec.Tolerations,
+		TerminationGracePeriod: &metav1.Duration{Duration: time.Duration(cluster.Spec.TerminationGracePeriodSecs)},
```

Comment:
We need to update the cockroach-operator module to the latest so that its Spec has latest configuration.